extensions-core/druid-lookups/src/main/java/io/druid/server/lookup/PollingLookup.java

```diff
@@ -0,0 +1,259 @@
+{
+  private static final Logger LOGGER = new Logger(PollingLookup.class);
+  private static final byte CACHE_ID = 0xa;
+
+  private final Period pollPeriod;
+
+  private final DataFetcher dataFetcher;
+
+  private final PollingCacheFactory cacheFactory;
+
+  private final AtomicReference<CacheRefKeeper> refOfCacheKeeper = new AtomicReference<>();
+
+  private final ListeningScheduledExecutorService scheduledExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newScheduledThreadPool(
+          1,
+          new ThreadFactoryBuilder()
+              .setDaemon(true)
+              .setNameFormat("PollingLookup-%d")
+              .setPriority(Thread.MIN_PRIORITY)
+              .build()
+      )
+  );
+
+  private final AtomicBoolean isOpen = new AtomicBoolean(false);
+
+  private final ListenableFuture<?> pollFuture;
+
+  public PollingLookup(
+      Period pollPeriod,
+      DataFetcher dataFetcher,
+      PollingCacheFactory cacheFactory
+  )
+  {
+    synchronized (isOpen) {
+      this.pollPeriod = pollPeriod == null ? Period.ZERO : pollPeriod;
+      this.dataFetcher = Preconditions.checkNotNull(dataFetcher);
+      this.cacheFactory = cacheFactory == null ? new OnHeapPollingCache.OnHeapPollingCacheProvider() : cacheFactory;
+      refOfCacheKeeper.set(new CacheRefKeeper(this.cacheFactory.makeOf(dataFetcher.fetchAll())));
+      if (getPollMs() > 0) {
+        pollFuture = scheduledExecutorService.scheduleAtFixedRate(
+            pollAndSwap(),
+            getPollMs(),
+            getPollMs(),
+            TimeUnit.MILLISECONDS
+        );
+      } else {
+        pollFuture = null;
+      }
+      this.isOpen.set(true);
+    }
+  }
+
+
+  public void close()
+  {
+    LOGGER.info("Closing lookup");
+    synchronized (isOpen) {
+      isOpen.getAndSet(false);
+      CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.getAndSet(null);
+      if (cacheRefKeeper != null) {
+        cacheRefKeeper.doneWithIt();
+      }
+      if (pollFuture != null) {
+        pollFuture.cancel(true);
+      }
+      scheduledExecutorService.shutdown();
+    }
+  }
+
+  @Override
+  public String apply(@NotNull String key)
+  {
+    assertIsOpen();
+    CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.get();
+    if (cacheRefKeeper == null) {
+      throw new ISE("Cache reference is null WTF");
+    }
+    PollingCache cache = cacheRefKeeper.get();
+    try {
+      if (cache == null) {
+        // it must've been closed while I was getting it.  Try again.
+        return this.apply(key);
+      }
+      return Strings.emptyToNull((String) cache.get(key));
+    }
+    finally {
+      if (cacheRefKeeper != null && cache != null) {
+        cacheRefKeeper.doneWithIt();
+      }
+    }
+  }
+
+  @Override
+  public List<String> unapply(final String value)
+  {
+    assertIsOpen();
+    CacheRefKeeper cacheRefKeeper = refOfCacheKeeper.get();
+    if (cacheRefKeeper == null) {
+      throw new ISE("lookup is closed");
+    }
```

Comment:
ok, interesting, that means assertIsOpen() check is not enough because closing can happen after that check. not sure what we buy from assertIsOpen() then?